cns/service/main.go

```diff
@@ -246,43 +248,90 @@ func printVersion() {
+		nodeRegisterRequest cns.NodeRegisterRequest
+	)
+
+	nodeRegisterRequest.NumCPU = numCPU
+	supportedApis, retErr := nmagentclient.GetNmAgentSupportedApis(httpc, "")
+
+	if retErr != nil {
+		logger.Errorf("[Azure CNS] Failed to retrieve SupportedApis from NMagent of node %s with Infrastructure Network: %s PrivateEndpoint: %s",
+			nodeID, infraVnet, dncEP)
+		return retErr
+	}
+
+	//To avoid any null-pointer deferencing errors.
+	if supportedApis == nil {
+		supportedApis = []string{}
+	}
+
+	nodeRegisterRequest.NmAgentSupportedApis = supportedApis
+
+	nodeRegisterTicker := time.NewTicker(acn.FiveSeconds)
+	responseChan := make(chan error)
+
+	for {
+		select {
+		case responseErr := <-responseChan:
+			return responseErr
+		case <-nodeRegisterTicker.C:
+			go sendRegisterNodeRequest(httpc, httpRestService, nodeRegisterRequest, url, responseChan)
+		}
+	}
+}
+
+// sendRegisterNodeRequest func helps in registering the node until there is an error.
+func sendRegisterNodeRequest(
+	httpc *http.Client,
+	httpRestService cns.HTTPService,
+	nodeRegisterRequest cns.NodeRegisterRequest,
+	registerURL string,
+	responseChan chan<- error) {
+
+	var (
+		body     bytes.Buffer
 		response *http.Response
 		err      = fmt.Errorf("")
-		body     bytes.Buffer
-		httpc    = acn.GetHttpClient()
 	)
 
-	for sleep := true; err != nil; sleep = true {
-		response, err = httpc.Post(url, "application/json", &body)
-		if err == nil {
-			if response.StatusCode == http.StatusCreated {
-				var req cns.SetOrchestratorTypeRequest
-				json.NewDecoder(response.Body).Decode(&req)
-				httpRestService.SetNodeOrchestrator(&req)
-				sleep = false
-			} else {
-				err = fmt.Errorf("[Azure CNS] Failed to register node with http status code %s", strconv.Itoa(response.StatusCode))
-				logger.Errorf(err.Error())
+	err = json.NewEncoder(&body).Encode(nodeRegisterRequest)
+	if err != nil {
+		log.Errorf("encoding json failed with %v", err)
+		responseChan <- err
+		return
+	}
+
+	response, err = httpc.Post(registerURL, "application/json", &body)
+	if err == nil {
+		if response.StatusCode == http.StatusCreated {
+			var req cns.SetOrchestratorTypeRequest
+			decodeErr := json.NewDecoder(response.Body).Decode(&req)
+			if decodeErr != nil {
+				log.Errorf("decoding Node Resgister response json failed with %v", err)
+				responseChan <- err
+				return
 			}
+			httpRestService.SetNodeOrchestrator(&req)
 
-			response.Body.Close()
+			logger.Printf("[Azure CNS] Node Registered")
+			responseChan <- nil
 		} else {
-			logger.Errorf("[Azure CNS] Failed to register node with err: %+v", err)
+			err = fmt.Errorf("[Azure CNS] Failed to register node with http status code %s", strconv.Itoa(response.StatusCode))
+			logger.Errorf(err.Error())
+			responseChan <- err
 		}
 
-		if sleep {
-			time.Sleep(acn.FiveSeconds)
-		}
+		response.Body.Close()
+	} else {
+		logger.Errorf("[Azure CNS] Failed to register node with err: %+v", err)
```

Comment:
Is there a reason why this path doesn't send the error? Also, i know this is just adding support of channels, but probably a good opportunity to clean up the function since its kinda hard to follow the code flow. Suggest returning early like:
```go
res, err := post()
if err != nil { /* return */ }

defer res.body.close()

if res.StatusCode != { /* return */ }

if err := decode(req); err != nil { /* return */ }

setNodeOrch(req)
```

Reply:
Okay will refactor. Yes, I am assuming this error is retriable error, like timeout or some other transport path error. (which most probably can be due to delays in setting up of azure services). The other errors are specific from DNC side, which i am attributing as not retriable.